internal/data/trace.go

```diff
@@ -0,0 +1,377 @@
+)
+
+// This file defines in-memory data structures to represent traces (spans).
+
+// ITraceData is the top-level struct that is propagated through the traces pipeline.
+// This is the newer version of consumerdata.TraceData, but uses more efficient
+// in-memory representation.
+type ITraceData struct {
+	resourceSpans []*ResourceSpans
+}
+
+func NewITraceData(resourceSpans []*ResourceSpans) ITraceData {
+	return ITraceData{resourceSpans}
+}
+
+// SpanCount calculates the total number of spans.
+func (td ITraceData) SpanCount() int {
+	spanCount := 0
+	for _, rsl := range td.resourceSpans {
+		spanCount += len(rsl.spans)
+	}
+	return spanCount
+}
+
+// A collection of spans from a Resource.
+type ResourceSpans struct {
+	// The resource for the spans in this message.
+	// If this field is not set then no resource info is known.
+	resource *Resource
+
+	// A list of Spans that originate from a resource.
+	spans []*Span
+}
+
+func NewResourceSpans(resource *Resource, spans []*Span) *ResourceSpans {
+	return &ResourceSpans{resource, spans}
+}
+
+func (m *ResourceSpans) Resource() *Resource {
+	return m.resource
+}
+
+func (m *ResourceSpans) SetResource(r *Resource) {
+	m.resource = r
+}
+
+func (m *ResourceSpans) Spans() []*Span {
+	return m.spans
+}
+
+func (m *ResourceSpans) SetSpans(s []*Span) {
+	m.spans = s
+}
+
+type TraceID struct {
+	bytes []byte
+}
+
+func TraceIDFromBytes(bytes []byte) TraceID { return TraceID{bytes} }
+
+type SpanID struct {
+	bytes []byte
+}
+
+func SpanIDFromBytes(bytes []byte) SpanID { return SpanID{bytes} }
+
+// TraceState in w3c-trace-context format: https://www.w3.org/TR/trace-context/#tracestate-header
+type TraceState string
+
+// Span represents a single operation within a trace. Spans can be
+// nested to form a trace tree. Spans may also be linked to other spans
+// from the same or different trace and form graphs. Often, a trace
+// contains a root span that describes the end-to-end latency, and one
+// or more subspans for its sub-operations. A trace can also contain
+// multiple root spans, or none at all. Spans do not need to be
+// contiguous - there may be gaps or overlaps between spans in a trace.
+// Must use NewSpan* functions to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type Span struct {
+	// Wrap OTLP Span.
+	orig *otlptrace.Span
+
+	// Override a few fields. These fields are the source of truth. Their counterparts
+	// stored in corresponding fields of "orig" are ignored.
+	attributes AttributesMap
+	events     []*SpanEvent
+	links      []*SpanLink
+}
+
+func NewSpan() *Span {
+	return &Span{orig: &otlptrace.Span{}}
+}
+
+// NewSpanSlice creates a slice of Spans that are correctly initialized.
+func NewSpanSlice(len int) []Span {
+	origs := make([]otlptrace.Span, len)
+	wrappers := make([]Span, len)
+	for i := range origs {
+		wrappers[i].orig = &origs[i]
+	}
+	return wrappers
+}
```

Comment:
Not sure I understand where is this used.

Reply:
See https://github.com/tigrannajaryan/opentelemetry-service/pull/1 for usage. It allows to avoid creating Spans one by one and saves a lot of allocations on large batches.